killswitch.py

```diff
@@ -1,54 +1,232 @@
 """Fetch kill switches from EDMC Repo."""
-from typing import Dict, List, NamedTuple, Optional, Union, cast
+from __future__ import annotations
+
+import threading
+from copy import deepcopy
+from typing import (
+    TYPE_CHECKING, Any, Callable, Dict, List, Mapping, MutableMapping, MutableSequence, NamedTuple, Optional, Sequence,
+    Tuple, TypedDict, TypeVar, Union, cast
+)
 
 import requests
 import semantic_version
+from semantic_version.base import Version
 
 import config
 import EDMCLogging
 
 logger = EDMCLogging.get_main_logger()
 
-DEFAULT_KILLSWITCH_URL = 'https://raw.githubusercontent.com/EDCD/EDMarketConnector/releases/killswitches.json'
-
+OLD_KILLSWITCH_URL = 'https://raw.githubusercontent.com/EDCD/EDMarketConnector/releases/killswitches.json'
+DEFAULT_KILLSWITCH_URL = 'https://raw.githubusercontent.com/EDCD/EDMarketConnector/releases/killswitches_v2.json'
+CURRENT_KILLSWITCH_VERSION = 2
+UPDATABLE_DATA = Union[Mapping, Sequence]
 _current_version: semantic_version.Version = config.appversion_nobuild()
 
+T = TypeVar('T', bound=UPDATABLE_DATA)
+
+
+class SingleKill(NamedTuple):
+    """A single KillSwitch. Possibly with additional rules."""
+
+    match: str
+    reason: str
+    redact_fields: Optional[List[str]] = None
+    delete_fields: Optional[List[str]] = None
+    set_fields: Optional[Dict[str, Any]] = None
+
+    @property
+    def has_rules(self) -> bool:
+        """Return whether or not this SingleKill can apply rules to a dict to make it safe to use."""
+        return any(x is not None for x in (self.redact_fields, self.delete_fields, self.set_fields))
+
+    def apply_rules(self, target: T) -> T:
+        """
+        Apply the rules this SingleKill instance has to make some data okay to send.
+
+        Note that this MODIFIES DATA IN PLACE.
+
+        :param target: data to apply a rule to
+        """
+        for key, value in (self.set_fields if self .set_fields is not None else {}).items():
+            _deep_apply(target, key, value)
+
+        for key in (self.redact_fields if self.redact_fields is not None else []):
+            _deep_apply(target, key, "REDACTED")
+
+        for key in (self.delete_fields if self.delete_fields is not None else []):
+            _deep_apply(target, key, delete=True)
+
+        return target
+
+
+def _apply(target: UPDATABLE_DATA, key: str, to_set: Any = None, delete: bool = False):
+    """
+    Set or delete the given target key on the given target.
+
+    :param target: The thing to set data on
+    :param key: the key or index to set the data to
+    :param to_set: the data to set, if any, defaults to None
+    :param delete: whether or not to delete the key or index, defaults to False
+    :raises ValueError: when an unexpected target type is passed
```

Comment:
```
NB: Other exceptions may be raised, i.e. IndexError if supplying a bad index.
```

My point here is to not have someone who's not read all the code/docs think that this can *only* raise ValueError.